src/apps/dialogs/TraitSelect/TraitSelectDialog.svelte

```diff
@@ -0,0 +1,263 @@
+    // updateDisabledStatus()
+    let _disabledTags = {};
+    let _disableSubmit = false;
+
+    // Sets the initial disable status
+    updateDisabledStatus();
+
+    function isCustomTrait(trait) {
+        return !traitGroups
+            .flatMap(({ traits }) => Object.keys(traits ?? {}))
+            .includes(trait);
+    }
+
+    function isDisabled(trait) {
+        // Disable all nonchecked, check marks if we've reached selectionLimit
+        if (
+            selectionLimit &&
+            selected.length >= selectionLimit &&
+            !selected.includes(trait)
+        ) {
+            return true;
+        }
+
+        return disabled.includes(trait);
+    }
+
+    function setTrait(trait, status) {
+        if (status) {
+            selected = [...selected, trait];
+        } else {
+            selected.findSplice((selected) => selected === trait);
+        }
+
+        // Prevent duplicate values
+        selected = [...new Set(selected)];
+        updateDisabledStatus();
+    }
+
+    function getCustomTrait() {
+        const custom = selected.filter(
+            (trait) => isCustomTrait(trait) && !disabled.includes(trait)
+        );
+
+        return custom.join(";");
+    }
+
+    function setCustomTrait(element) {
+        // Split, trim, remove empty & disabled
+        const customList = element.value
+            .split(";")
+            .map((i) => i.trim())
+            .filter((i) => i && !disabled.includes(i));
+
+        selected = selected
+            // Remove all custom from selected
+            .filter((trait) => !isCustomTrait(trait))
+            // Add back disabled selected (our input box doesn't control these)
+            .concat(_disabledCustomSelectedTraits);
+
+        // Test to ensure that the custom added selection isn't going to exceed the limit.
+        if (selectionLimit) {
+            // Calculate how much we're over the limit by.
+            const overLimit =
+                selected.length + customList.length - selectionLimit;
+
+            // Warn the user and truncate the customList
+            if (overLimit > 0) {
+                ui.notifications.warn(
+                    "You have added too many custom traits, we will remove the extra for you"
+                );
+                customList.splice(-overLimit);
+            }
+        }
+
+        // Add the custom traits to the selected and update
+        selected = [...selected, ...customList];
+        element.value = getCustomTrait();
+        updateDisabledStatus();
+    }
+
+    function submitForm() {
+        application.submit(selected);
+    }
+
+    // Updates the disabled/checked values. Since any checkbox can modifiy the disabled state of any
+    // Due to the selectionLimit, we must maintain the list of _disabled elemets seperately so that
+    // Svelte knows to update those elements.
+    function updateDisabledStatus() {
+        for (const group of traitGroups) {
```

Comment:
Not too sure if double for loop is the way to go, but the alternative is kinda still meh cz of the `_disabledTags[traits] = ...` portion,